### ISSUE TITLE
Importing/Exporting JsonLogic with additional operators

### ISSUE THREAD

**Issue by hwhmeikle** [react-querybuilder/react-querybuilder#398]
Firstly, thank you very much for this library. It's been a breeze to use!

I'm in the process of replacing `react-awesome-query-builder` with this package and have hit my first hitch.

We are using JsonLogic with some additional operators such as:
`JSONLogic.add_operation("lessThanOrEqualDate", isBeforeOrEqual);`

When attempting to import a query that uses this operator with `parseJsonLogic`  e.g.
```
{
  "query": {
    "and": [
      {
        "lessThanOrEqualDate": [
          {
            "var": "device.authenticatedAt"
          },
          "PT-24H"
        ]
      }
    ]
  }
}
```

Then passing it into `parseJsonLogic` it seems to strip it out. However, any condition that is using one of the `defaultOperators` works fine.

I assume it's something to do with `parseJsonLogic` being scoped to the default operators. Is there anyway to let it recognise additional ones?

**Comment by jakeboone02** [comment]
Hey @hwhmeikle, thanks for the kind words! I'm curious about what's driving your migration from RAQB to RQB. Any feedback is appreciated!

The `parse*` functions bail out (skip the rule) when they don't know what to do, and that includes rules with non-standard operators. This could be addressed in userland (i.e. you provide your own workaround) or with a configuration to instruct the function how to handle unknown operators.

I'd rather try the former option first. Could you transform the JsonLogic object to map any new operators to known operators first, and then parse it? Your example would change like this, I think:

```diff
 parseJsonLogic({
   and: [
     {
-      lessThanOrEqualDate: [
+      "<=": [
         { var: "device.authenticatedAt" },
         "PT-24H"
       ]
     }
   ]
 })
```

If that doesn't work, we'd probably need to go with the latter option and create a configuration, maybe a map from custom operators to standard ones so `parseJsonLogic` knows what to do with each one. Let me know what you think.

**Comment by hwhmeikle** [comment]
Thanks for quick reply! The main reason for the migration is that whilst RAQB is very powerful, the DX feels cumbersome once you need to add any custom behaviour. Our main use case is allowing users to create queries and essentially create their own custom fields on the fly. I won't go into detail but the process of updating the config/fields in RAQB and not having it break was fairly hacky and not reacty at all.

I'll give what you suggested a go, and let you know - cheers!

**Comment by hwhmeikle** [comment]
Update: I've written the logic for transforming the keys for importing it and that's working well!

My next hurdle is transforming the other way e.g. `<=` -> `lessThanOrEqualDate`. I only want to transform `<=` operator to `lessThanOrEqualDate` when it is for the `device.authenticatedAt` field.

Would there be a way to use `transformQuery` here to achieve this, or is my best bet writing my own custom transformer?

**Comment by jakeboone02** [comment]
I don't think you need to use `transformQuery`, just pass a custom `ruleProcessor` to `formatQuery` like below ([CodeSandbox here](https://codesandbox.io/s/friendly-fog-5o36e4?file=/src/App.tsx)):

```ts
import {
  defaultRuleProcessorJsonLogic,
  formatQuery,
  RuleGroupType,
  RuleType
} from "react-querybuilder";

const query: RuleGroupType = {
  combinator: "and",
  rules: [{ field: "device.authenticatedAt", operator: "<=", value: "PT-24H" }]
};

const jsonlgc = formatQuery(query, {
  format: "jsonlogic",
  ruleProcessor: (r: RuleType) => {
    if (r.field === "device.authenticatedAt") {
      return { lessThanOrEqualDate: defaultRuleProcessorJsonLogic(r)["<="] };
    }
    return defaultRuleProcessorJsonLogic(r);
  }
});
```

What that's doing is taking `'<='` property from the result of the default rule processor and assigning it to the `lessThanOrEqualDate` property of a new object (unless you're dealing with a different `field`, in which case it just falls back to the default rule processor).

You may need to add some error checking and other contingencies but hopefully that gets you headed in the right direction.

**Comment by hwhmeikle** [comment]
@jakeboone02 Thank you that works nicely :).

You've already gone above and beyond, but would be keen to hear your thoughts on approaching converting the `multiselect` output from RAQB to RQB.

When using a `multiselect` in RAQB it seems to wrap the entire rule with an `all` operator e.g.
```
{
  "and": [
    {
      "all": [
        {
          "var": "multicolor"
        },
        {
          "in": [
            {
              "var": ""
            },
            [
              "green",
              "yellow",
              "orange"
            ]
          ]
        }
      ]
    }
  ]
}
```

When using a `multiselect` in RQB it doesn't wrap it in an `all` operator (which to me makes more sense) e.g. 

```
{
  "in": [
    {
      "var": "multicolor"
    },
    [
      "green",
      "yellow",
      "orange"
    ]
  ]
}
```

My thinking is that I need to somehow transform the RAQB output to match the RQB output as RQB doesn't use `all` internally so the rule gets skipped when using `parseJsonLogic`.

**Comment by jakeboone02** [comment]
That's confusing to me. I _think_ RAQB is not doing what one would typically expect, but maybe I'm misreading it. For example, this RAQB rule:

![image](https://user-images.githubusercontent.com/366438/201778617-5634bc96-4060-4cef-b5be-76dddc766eee.png)

generates this JsonLogic rule:

```ts
{ "and": [{ "all": [{ "var": "multicolor" }, { "in": [{ "var": "" }, [ "yellow", "orange" ]] }] }] }
```

But that rule passes when the data is `{ "multicolor": ["yellow"] }` (notice the value is an array). It's saying "is the value an array that contains a value in this other array?" I would expect the rule to pass instead for `{ "multicolor": "yellow" }`, which would say (more simply, and I believe more intuitively) "is the value in this array?" That would also be more like the SQL `IN` operator.

So I'm not sure there actually is a direct translation from the RAQB output to what RQB handles, at least not that keeps the same logic. Like I said, maybe I'm misreading something.

**Comment by hwhmeikle** [comment]
@jakeboone02 I'm on the same page and would expect it to work as you describe. We've had to get around it by turning the value into an array as you've shown. I guess keeping the same logic would mean turning `["yellow"]` -> `"yellow"` when importing.

**Comment by jakeboone02** [comment]
Yeah that would make more sense to me. You'll still have to translate the "all" operator. Looks like you should take the first element of the `"all"` array and use that as the first element of the `"in"` array instead of `{ "var": "" }`. Then extract the `"in"` rule and process that instead of the full `"all"` rule.

**Comment by hwhmeikle** [comment]
@jakeboone02 Thanks for the help! Will close this issue.

**Comment by jakeboone02** [comment]
@hwhmeikle I just released [v5.3.2](https://github.com/react-querybuilder/react-querybuilder/releases/tag/v5.3.2) with a new configuration option for `parseJsonLogic` called `jsonLogicOperations`. Documentation [here](https://react-querybuilder.js.org/docs/api/import#custom-operations).

You can now specify a function that returns a rule or group based on any custom operation.
